File: 2015/01/2.js
"use strict";
const fs = require("node:fs");
let contents = fs.readFileSync("./2015/01/input.txt", "utf8").toString();

let floor = 0;
let count = 0;
for (let character of contents) {
  count++;
  if (character == "(") {
    floor++;
  } else if (character == ")") {
    floor--;
  } else {
    console.warn(`Invalid character '${character}' in input file`);
  }
  if (floor == -1) {
    console.log(count);
    process.exit(0);
  }
}

console.error("Santa will never reach the basement");

File: 2015/01/2.c
#include <stdio.h>
#include <stdlib.h>

#define basement_floor -1

int main(void) {
  FILE *input = fopen("./2015/01/input.txt", "r");
  if (!input) {
    fprintf(stderr, "Unable to open input file\n");
    return EXIT_FAILURE;
  }
  int floor = 0;
  int c;
  unsigned int iteration = 0;
  while ((c = fgetc(input)) != EOF) {
    iteration++;
    switch (c) {
    case '(':
      floor++;
      break;
    case ')':
      floor--;
      break;
    default:
      fprintf(stderr, "Warning invalid character: '%c'\n", c);
      break;
    }
    if (floor == basement_floor) {
      printf("%d", iteration);
      return EXIT_SUCCESS;
    }
  }
  fclose(input);
  fprintf(stderr, "Santa never reached the basement");
  return EXIT_FAILURE;
}
